Add ADM Mass / Spin and Extrapolation to Infinity

TaigoFr opened the pull request:
Addition of a class `ADMMass` that calculates the ADM mass and spin of a spacetime, together with a class `ADMMassExtraction` that extracts the corresponding variables.

This partially solves #90 , and will have to be changed once diagnostics are added. #90 also includes the Kretschmann scalar and such variables, but that's in my opinion a feature to be added separately.

This PR also includes Extrapolation to infinity of first and second order. First order is in Alcubierre. Second order I wrote some notes about it, as I didn't find it elsewhere and made the computation myself. Second order (requires 3 radii to cancel leading terms of order 1/r AND 1/r^2) greatly improves the result, so it must be correct.

I added this extraction to the Kerr example, as it's a neat place for it to be. For a Kerr BH with `Mass=1.0` and `Spin=0.3` I managed to get in my laptop a quick extraction that with 2nd order extraction to infinity results in `M=1.0000609955` and `a=0.29994697676`, which looks pretty good to me. I used `N=64`, `L=128`, extraction radii at `r = (30. 40. 50.)` at levels `(1 1 0)`. The simpson method of @mirenradia's SurfaceExtraction also greatly helps this. Other methods (except for boole rule) or first order extraction get slightly worse results, but still converge.

ADM extraction has some ambiguity in terms of what factors of chi should be added. In principle it doesn't matter. I tested that indeed it doesn't matter much, once extrapolation is done. So I added them trying to follow standard calculations, but it is open for review.

KAClough commented:
This is also very nice! Especially nice to have the extrapolation of the quantities to infinity output automatically. 

A few minor comments which I will put in the code.

KAClough commented on Examples/KerrBH/KerrBHLevel.cpp:
```diff
@@ -84,3 +87,26 @@ void KerrBHLevel::computeTaggingCriterion(FArrayBox &tagging_criterion,
 {
     BoxLoops::loop(ChiTaggingCriterion(m_dx), current_state, tagging_criterion);
```
We should probably use the extraction tagging criterion to make sure we have regridded to the right level, even though we only use the low levels for this case.

KAClough commented on Examples/KerrBH/KerrBHLevel.cpp:
```diff
@@ -84,3 +87,26 @@ void KerrBHLevel::computeTaggingCriterion(FArrayBox &tagging_criterion,
+    if (m_p.activate_extraction == 1)
+    {
+        // Populate the Weyl Scalar values on the grid
+        fillAllGhosts();
```
Comment should say Madm not Weyl. 

I have always thought that filling the Weyl/Madm data on every level post timestep is quite inefficient. This is especially true here when we only actually use it every coarse or second to coarse timestep. 

Perhaps we could do something like
if(m_time % extraction_dt == 0) {calculate diagnostics}
I think the modulo thing is usually non trivial as you have to check to numerical precision. Having a small function in GRAMRLevel that returns true if m_time is a multiple of dt for a given level would be quite useful generally, and could then be used here. 

This can be filed as an issue if you don't want to do it now :-)

TaigoFr replied:
A quick alternative is to just move it to inside the following `if(m_level == m_p.extraction_params.min_extraction_level())` no? Like that it is only calculated on the moments we want output. I'm not sure this is the same as what you wanted, let me know.

mirenradia replied:
Careful @TaigoFr, you want the `BoxLoops::loop` called on all levels, not just the extraction level.

TaigoFr replied:
I realized that and edited my comment. I solved it in the last commit. It's a solution that is stable under numerical perturbations.

KAClough replied:
Nice fix :-)
Can we also add it to BinaryBHLevel::specificPostTimeStep()? It probably doesn't save us much, but it annoys me that we calculate Weyl on every level timestep.

KAClough commented on Examples/KerrBH/params_cheap.txt:
```diff
@@ -0,0 +1,108 @@
+sigma = 0.3
+
+#extraction params
+activate_extraction = 1
```
As in the other pull request, we should update params.txt and make params_cheap.txt more trivial.

TaigoFr replied:
Yeah I might have to fix these parameters either here or on #120 , depending on which is merged first. I'll change a few things. Not sure if params.txt will be good. params_cheap.txt I'm not sure what to do with them. I think I might want to change them again anyway for the AHFinder PR.

KAClough replied:
Thats fine, we can amend them later once you know which values are best for testing. The changes you made look fine to me.

KAClough commented on Source/CCZ4/ADMMass.hpp:
```diff
@@ -0,0 +1,126 @@
+    const std::array<double, CH_SPACEDIM> &m_center;
+    const double m_G_Newton; //!< Newton's constant
+
+    DIR dir;
```
dir -> m_dir so it can be easily recognised as a member variable.

KAClough commented on Source/utils/ADMMassExtraction.hpp:
```diff
@@ -0,0 +1,77 @@
+
+        if (m_params.write_extraction)
+        {
+            write_extraction("Weyl4ExtractionOut_");
```
Find and replace Weyl -> Madm :-)

KAClough commented:
Subject to the small change I suggested for amending the BBH example I'm happy with this, @mirenradia are you?

mirenradia commented:
I will make a full review at some point but my main comment at the moment is that I think the code for ADM angular momentum should be separated out from the code for the ADM mass (and their respective extraction classes) and `ChomboParameters::variable_names_to_enum` shouldn't be used to check for this (especially in a `compute` class where performance is important).

mirenradia commented:
Now that #119 is merged, we should make these diagnostic variables.

KAClough commented:
I am wondering pedantically about the "ADMMass" name. It also includes the angular momentum, so I think that as Miren suggests this could be separated out, or it could become ADMQuantities or ADMMetrics or some other name which makes it sound more general. I would also like to add the ADM Momentum in some specified direction (doesn't have to be now, but maybe this influences how we solve this issue).

Apart from this (and potential merge conflicts with the other PRs) I think it looks great.

TaigoFr commented:
I changed the name from ADMMass to ADMQuantities. I think it's best to not separate the mass and momentum calculations, as they involve the common calculation of the normal `dS`. Since now the mass and momentum diagnostic components are specified in the constructor and defaulted to "don't calculate" (`a_c_Madm = -1, int a_c_Jadm = -1`), it's hopefully good enough.
Updated to the most recent code as well.

mirenradia commented:
Can it be made possible for the user to select (via a parameter) which radii to use for the extrapolation?

mirenradia commented on Source/CCZ4/ADMQuantities.hpp:
```diff
@@ -0,0 +1,132 @@
+            data_t Madm = 0.0;
+            FOR(i, j, k, l)
+            {
+                Madm += dS_L[i] / (16. * M_PI * m_G_Newton) /
+                        (vars.chi * sqrt(vars.chi)) * h_UU[j][k] * h_UU[i][l] *
+                        (vars.chi * (d1.h[l][k][j] - d1.h[j][k][l]) -
+                         (vars.h[l][k] * d1.chi[j] - vars.h[j][k] * d1.chi[l]));
```
Maybe add a reference for these formulae as a comment e.g. Baumgarte and Shapiro p85 Eq. (3.128)?

mirenradia commented on Source/CCZ4/ADMQuantities.hpp:
```diff
@@ -0,0 +1,132 @@
+
+            FOR(i, j, k)
+            {
+                Jadm += -dS_L[i] / (8. * M_PI * m_G_Newton) *
+                        epsilon[m_dir][j][k] * x[j] * vars.K *
+                        TensorAlgebra::delta(i, k);
+
+                FOR(l, m)
+                {
+                    Jadm += dS_L[i] / (8. * M_PI * m_G_Newton) *
+                            epsilon[m_dir][j][k] * x[j] * h_UU[i][l] *
+                            h_UU[k][m] * vars.chi *
+                            (vars.A[l][m] + vars.K * vars.h[l][m] / 3.);
```
Add a reference e.g. Baumgarte and Shapiro p95 Eq. (3.191).

mirenradia reviewed (changes requested):
Thanks for adding this code and sorry for the long delay in reviewing.

Would it be possible to add the Richardson extrapolation in a new `write_xxx` function e.g. `SurfaceExtraction::write_integrals_and_extrapolate` rather than hacking it into the existing function?
